lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_facts.py

```diff
@@ -0,0 +1,422 @@
+                "name": "OmsAgentForLinux",
+                "properties": {
+                    "autoUpgradeMinorVersion": true,
+                    "provisioningState": "Succeeded",
+                    "publisher": "Microsoft.EnterpriseCloud.Monitoring",
+                    "settings": {
+                        "azureResourceId": "/subscriptions/XXXXX/resourcegroups/Testing/providers/microsoft.compute/virtualmachines/vm",
+                        "stopOnMultipleConnections": true,
+                        "workspaceId": "43ac2095-42c7-4488-924f-bc4cbca63f8d"
+                    },
+                    "type": "OmsAgentForLinux",
+                    "typeHandlerVersion": "1.0"
+                },
+                "type": "Microsoft.Compute/virtualMachines/extensions"
+            }
+        ],
+        "tags": {
+            "cluster_name": "cluster"
+        },
+        "type": "Microsoft.Compute/virtualMachines"
+    }]
+'''
+
+try:
+    from msrestazure.azure_exceptions import CloudError
+except:
+    # This is handled in azure_rm_common
+    pass
+
+from ansible.module_utils.azure_rm_common import AzureRMModuleBase, azure_id_to_dict
+
+
+AZURE_OBJECT_CLASS = 'VirtualMachine'
+
+AZURE_ENUM_MODULES = ['azure.mgmt.compute.models']
+
+
+class AzureRMVirtualMachineFacts(AzureRMModuleBase):
+
+    def __init__(self):
+
+        self.module_arg_spec = dict(
+            resource_group=dict(type='str'),
+            name=dict(type='str'),
+            tags=dict(type='list'),
+            auth_source=dict(type='str', choices=['auto', 'cli', 'env', 'credential_file', 'msi'], default='auto'),
+            cloud_environment=dict(type='str', default='AzureCloud')
+        )
+
+        self.results = dict(
+            changed=False,
+            ansible_facts=dict(azure_virtualmachines=[])
+        )
+
+        self.resource_group = None
+        self.name = None
+        self.tags = None
+
+        super(AzureRMVirtualMachineFacts, self).__init__(self.module_arg_spec,
+                                                         supports_tags=False,
+                                                         facts_module=True)
+
+    def exec_module(self, **kwargs):
+
+        for key in self.module_arg_spec:
+            setattr(self, key, kwargs[key])
+
+        if self.name and not self.resource_group:
+            self.fail("Parameter error: resource group required when filtering by name.")
+        if self.name:
+            self.results['ansible_facts']['azure_virtualmachines'] = self.get_item()
+        else:
+            self.results['ansible_facts']['azure_virtualmachines'] = self.list_items()
+
+        return self.results
+
+    def get_item(self):
+        self.log('Get properties for {0}'.format(self.name))
+        item = None
+        result = []
+
+        try:
+            item = self.compute_client.virtual_machines.get(self.resource_group, self.name)
+        except CloudError:
+            pass
```

Comment:
We probably want to log the exception as a warning if possible and not just ignore it

Reply:
Will be changed for the next commit. The warning message will be like the example:

```txt
TASK [Get the resource group VMs facts] **********************************************************************************************************************
 [WARNING]: Error getting virtual machine confluence1 - Azure Error: ResourceNotFound Message: The Resource 'Microsoft.Compute/virtualMachines/confluence1'
under resource group 'Dev_Devops1' was not found.
```